lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp

```diff
@@ -1777,6 +1777,162 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+                axes, zero, selectIndex);
+            Value dim = rewriter.create<Torch::AtenItemOp>(binder.getLoc(),
+                                                           torchIntTy, extract);
+            axesList.push_back(dim);
+          }
+        }
+
+        // Handle the noop case:
+        // When axes is empty and noop_with_empty_axes is set to true, input
+        // tensor will not be reduced, and the output tensor would be
+        // equivalent to input tensor.
+        if (axesList.empty() && noop_with_empty_axes) {
+          rewriter.replaceOp(binder.op, data);
+          return success();
+        }
+
+        // Handle case when no axes arg is passed but not a noop:
+        // Manually set positive axis to all dims.
+        if (axesList.empty()) {
+          for (int i = 0; i < rank; i++) {
+            Value dimValue = rewriter.create<Torch::ConstantIntOp>(
+                binder.getLoc(), rewriter.getI64IntegerAttr(i));
+            axesList.push_back(dimValue);
+          }
+        }
+
+        // Handle negative axis:
+        Value rankVal = rewriter.create<Torch::AtenDimOp>(binder.getLoc(),
+                                                          torchIntTy, data);
+        for (Value &axes : axesList) {
+          Value isNegative =
+              rewriter.create<Torch::AtenLtIntOp>(binder.getLoc(), axes, zero);
+          isNegative = rewriter.create<Torch::AtenIntBoolOp>(binder.getLoc(),
+                                                             isNegative);
+          Value finalOffset = rewriter.create<Torch::AtenMulIntOp>(
+              binder.getLoc(), isNegative, rankVal);
+          axes = rewriter.create<Torch::AtenAddIntOp>(binder.getLoc(), axes,
+                                                      finalOffset);
+        }
+
+        // Handle multiple axes case:
+        // ReduceProd on each dim, always set keepDimsBool == True to avoid
+        // segfault.
+        Value trueVal =
+            rewriter.create<Torch::ConstantBoolOp>(binder.getLoc(), true);
+        Value noneVal = rewriter.create<Torch::ConstantNoneOp>(binder.getLoc());
+        SmallVector<int64_t> intermediateShape;
+        for (int i = 0; i < rank; i++) {
+          intermediateShape.push_back(Torch::kUnknownSize);
+        }
+        Value dataReduceProd = data;
+        for (auto axis : axesList) {
+          Type resultTyReduceProd = dataTy.getWithSizesAndDtype(
+              ArrayRef(intermediateShape), dataTy.getOptionalDtype());
+          dataReduceProd = rewriter.create<Torch::AtenProdDimIntOp>(
+              binder.getLoc(), resultTyReduceProd, dataReduceProd, axis,
+              trueVal, noneVal);
+        }
+
+        // Derived the final shape of the tensor after prod loop of each axis.
+        SmallVector<int64_t> dataReduceProdSize;
+        auto dataSize = dataTy.getSizes();
+        auto resultTypeSizes = resultType.getSizes();
+        if (!keepDims) {
```

Comment:
You don't need an if / else if keep dims is true. Instead you can just do
```
if (keepdims) {
  rewriter.replaceOp(binder.op, dataReduceProd);
  return success();
}
```

Then continue with the additional reshape.

Reply:
We cannot directly return the dataReduceProd. Its shape is unknown after the loop. So the reshape is necessary. 
If we do so, it will output error mlir like this.
```
func.func @test_reduce_prod_default_axes_keepdims_random(%arg0: !torch.vtensor<[3,2,2],f32>) -> !torch.vtensor<[1,1,1],f32> attributes {torch.onnx_meta.ir_version = 8 : si64, torch.onnx_meta.opset_version = 18 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
  %int0 = torch.constant.int 0
  %int0_0 = torch.constant.int 0
  %int1 = torch.constant.int 1
  %int2 = torch.constant.int 2
  %0 = torch.aten.dim %arg0 : !torch.vtensor<[3,2,2],f32> -> !torch.int
  %1 = torch.aten.lt.int %int0_0, %int0 : !torch.int, !torch.int -> !torch.bool
  %2 = torch.aten.Int.bool %1 : !torch.bool -> !torch.int
  %3 = torch.aten.mul.int %2, %0 : !torch.int, !torch.int -> !torch.int
  %4 = torch.aten.add.int %int0_0, %3 : !torch.int, !torch.int -> !torch.int
  %5 = torch.aten.lt.int %int1, %int0 : !torch.int, !torch.int -> !torch.bool
  %6 = torch.aten.Int.bool %5 : !torch.bool -> !torch.int
  %7 = torch.aten.mul.int %6, %0 : !torch.int, !torch.int -> !torch.int
  %8 = torch.aten.add.int %int1, %7 : !torch.int, !torch.int -> !torch.int
  %9 = torch.aten.lt.int %int2, %int0 : !torch.int, !torch.int -> !torch.bool
  %10 = torch.aten.Int.bool %9 : !torch.bool -> !torch.int
  %11 = torch.aten.mul.int %10, %0 : !torch.int, !torch.int -> !torch.int
  %12 = torch.aten.add.int %int2, %11 : !torch.int, !torch.int -> !torch.int
  %true = torch.constant.bool true
  %none = torch.constant.none
  %13 = torch.aten.prod.dim_int %arg0, %4, %true, %none : !torch.vtensor<[3,2,2],f32>, !torch.int, !torch.bool, !torch.none -> !torch.vtensor<[?,?,?],f32>
  %14 = torch.aten.prod.dim_int %13, %8, %true, %none : !torch.vtensor<[?,?,?],f32>, !torch.int, !torch.bool, !torch.none -> !torch.vtensor<[?,?,?],f32>
  %15 = torch.aten.prod.dim_int %14, %12, %true, %none : !torch.vtensor<[?,?,?],f32>, !torch.int, !torch.bool, !torch.none -> !torch.vtensor<[?,?,?],f32>
  %16 = torch.operator "onnx.ReduceProd"(%arg0) {torch.onnx.keepdims = 1 : si64} : (!torch.vtensor<[3,2,2],f32>) -> !torch.vtensor<[1,1,1],f32> 
  return %16 : !torch.vtensor<[1,1,1],f32>
}
```
`error: failed to materialize conversion for result #0 of operation 'torch.operator' that remained live after conversion`

Reply:
You can work around that by using resultType as the final result type on the AtenProdDimInt correct?